### PR TITLE
Add playground to simulate transfers with varying parameters

### PR DESCRIPTION
**Pull request by hadjiszs**
I'm trying to do some playground with at2 through your implementation.
I wanna try to run in different configurations with evolving value parameters, that's why I went for a rust example rather than the tests.

However, I got some issues in my way that seems to be related to the sequences with sieve.

Here is the relevant part of the logs:
```
Running `target/debug/examples/double_spend`
2021-11-16T11:05:51.889006Z  INFO double_spend: Launching network of 4 nodes
2021-11-16T11:05:52.668597Z  INFO double_spend: Network of 4 nodes launched
2021-11-16T11:05:52.668651Z  INFO double_spend: addr="http://127.0.0.1:3000/" rpc="http://127.0.0.1:3001/"
2021-11-16T11:05:52.668684Z  INFO double_spend: addr="http://127.0.0.1:3002/" rpc="http://127.0.0.1:3003/"
2021-11-16T11:05:52.668719Z  INFO double_spend: addr="http://127.0.0.1:3004/" rpc="http://127.0.0.1:3005/"
2021-11-16T11:05:52.668750Z  INFO double_spend: addr="http://127.0.0.1:3006/" rpc="http://127.0.0.1:3007/"
2021-11-16T11:05:52.697270Z  INFO double_spend: Out of a network of 4 nodes: node_i=3 node_k=2
2021-11-16T11:05:52.697298Z  INFO double_spend: Account 0 is sending to account 2
2021-11-16T11:05:52.715391Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3007/"
2021-11-16T11:05:52.715423Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-16T11:06:02.743992Z  INFO double_spend: Transfer 1/10 done
2021-11-16T11:06:02.744066Z  INFO double_spend: Out of a network of 4 nodes: node_i=1 node_k=2
2021-11-16T11:06:02.744081Z  INFO double_spend: Account 5 is sending to account 7
2021-11-16T11:06:02.767347Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-16T11:06:02.767384Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3005/"
thread 'main' panicked at 'sender: 100 - 1 != 100', examples/double_spend.rs:306:5
```

Do you have any clue of why the second tx does not succeed even though it involves two different accounts and two different servers?

Thank you!

### CODE REVIEW

File: examples/double_spend.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+use futures::future::join_all;
+//use futures::executor::block_on;
+use tokio::{net::TcpStream, task::yield_now};
+use tracing::info;
+use tracing_subscriber;
+use structopt::StructOpt;
+use rand::{thread_rng, Rng};
+
+const CLIENT_BIN: &str = "/home/mhadji/research/at2/at2-node/target/debug/client";
+const SERVER_BIN: &str = "/home/mhadji/research/at2/at2-node/target/debug/server";
```
Comment (hadjiszs) [inline examples/double_spend.rs]: Sorry for the absolute paths, `CARGO_BIN_EXE_<name>` seems to be not exposed for the rust examples

**Comment by tharvik** [comment]
> Do you have any clue of why the second tx does not succeed even though it involves two different accounts and two different servers?

You're right, there is an issue when processing blocks from the same node. I [fixed it](https://github.com/Distributed-EPFL/murmur/pull/14) in `murmur`, so you should be good to go after a `cargo update` :)

> However, I got some issues in my way that seems to be related to the sequences with sieve.

I see that in `generate_transfer:transfer` you do not increment the sequence number, you should do so for each new transaction from a sender. For now, you need to start at `1` (as the first sequence is supposedly used to top up the account).

After doing that, your example indeed works
```
2021-11-17T15:30:09.907652Z  INFO double_spend: Launching network of 4 nodes
2021-11-17T15:30:10.599153Z  INFO double_spend: Network of 4 nodes launched
2021-11-17T15:30:10.599231Z  INFO double_spend: addr="http://127.0.0.1:3000/" rpc="http://127.0.0.1:3001/"
2021-11-17T15:30:10.599262Z  INFO double_spend: addr="http://127.0.0.1:3002/" rpc="http://127.0.0.1:3003/"
2021-11-17T15:30:10.599285Z  INFO double_spend: addr="http://127.0.0.1:3004/" rpc="http://127.0.0.1:3005/"
2021-11-17T15:30:10.599306Z  INFO double_spend: addr="http://127.0.0.1:3006/" rpc="http://127.0.0.1:3007/"
2021-11-17T15:30:10.628501Z  INFO double_spend: Out of a network of 4 nodes: node_i=0 node_k=1
2021-11-17T15:30:10.628535Z  INFO double_spend: Account 1 is sending to account 3
2021-11-17T15:30:10.649737Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3001/"
2021-11-17T15:30:10.649768Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-17T15:30:20.675072Z  INFO double_spend: Transfer 1/10 done
2021-11-17T15:30:20.675185Z  INFO double_spend: Out of a network of 4 nodes: node_i=2 node_k=3
2021-11-17T15:30:20.675205Z  INFO double_spend: Account 2 is sending to account 4
2021-11-17T15:30:20.701019Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-17T15:30:20.701048Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3007/"
2021-11-17T15:30:30.728209Z  INFO double_spend: Transfer 2/10 done
2021-11-17T15:30:30.728288Z  INFO double_spend: Out of a network of 4 nodes: node_i=1 node_k=2
2021-11-17T15:30:30.728320Z  INFO double_spend: Account 8 is sending to account 6
2021-11-17T15:30:30.749804Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-17T15:30:30.749833Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-17T15:30:40.776302Z  INFO double_spend: Transfer 3/10 done
2021-11-17T15:30:40.776371Z  INFO double_spend: Out of a network of 4 nodes: node_i=3 node_k=2
2021-11-17T15:30:40.776402Z  INFO double_spend: Account 5 is sending to account 7
2021-11-17T15:30:40.797121Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3007/"
2021-11-17T15:30:40.797150Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-17T15:30:50.821999Z  INFO double_spend: Transfer 4/10 done
2021-11-17T15:30:50.822100Z  INFO double_spend: Out of a network of 4 nodes: node_i=0 node_k=1
2021-11-17T15:30:50.822118Z  INFO double_spend: Account 2 is sending to account 4
2021-11-17T15:30:50.843144Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3001/"
2021-11-17T15:30:50.843171Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-17T15:31:00.868334Z  INFO double_spend: Transfer 5/10 done
2021-11-17T15:31:00.868411Z  INFO double_spend: Out of a network of 4 nodes: node_i=2 node_k=3
2021-11-17T15:31:00.868454Z  INFO double_spend: Account 0 is sending to account 2
2021-11-17T15:31:00.888757Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-17T15:31:00.888783Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3007/"
2021-11-17T15:31:10.916730Z  INFO double_spend: Transfer 6/10 done
2021-11-17T15:31:10.916845Z  INFO double_spend: Out of a network of 4 nodes: node_i=1 node_k=2
2021-11-17T15:31:10.916875Z  INFO double_spend: Account 0 is sending to account 2
2021-11-17T15:31:10.939140Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-17T15:31:10.939171Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-17T15:31:20.964533Z  INFO double_spend: Transfer 7/10 done
2021-11-17T15:31:20.964615Z  INFO double_spend: Out of a network of 4 nodes: node_i=0 node_k=1
2021-11-17T15:31:20.964628Z  INFO double_spend: Account 1 is sending to account 3
2021-11-17T15:31:20.985921Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3001/"
2021-11-17T15:31:20.985949Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-17T15:31:31.010608Z  INFO double_spend: Transfer 8/10 done
2021-11-17T15:31:31.010677Z  INFO double_spend: Out of a network of 4 nodes: node_i=0 node_k=1
2021-11-17T15:31:31.010707Z  INFO double_spend: Account 6 is sending to account 4
2021-11-17T15:31:31.031497Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3001/"
2021-11-17T15:31:31.031526Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3003/"
2021-11-17T15:31:41.056345Z  INFO double_spend: Transfer 9/10 done
2021-11-17T15:31:41.056469Z  INFO double_spend: Out of a network of 4 nodes: node_i=3 node_k=2
2021-11-17T15:31:41.056497Z  INFO double_spend: Account 7 is sending to account 5
2021-11-17T15:31:41.077943Z  INFO double_spend: sender-config: rpc_address = "http://127.0.0.1:3007/"
2021-11-17T15:31:41.077977Z  INFO double_spend: receiver-config: rpc_address = "http://127.0.0.1:3005/"
2021-11-17T15:31:51.109850Z  INFO double_spend: Transfer 10/10 done
2021-11-17T15:31:51.109895Z  INFO double_spend: Shutting down the network
```

File: examples/double_spend.rs (hunk 1 above)
Comment (tharvik) [reply examples/double_spend.rs]: Indeed, [it's "only set when building an integration test or benchmark"](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates). You can however use another variable, `CARGO_MANIFEST_DIR`, which is the root of the crate.

```suggestion
const CLIENT_BIN: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/target/debug/client");
const SERVER_BIN: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/target/debug/server");
```

Comment (hadjiszs) [reply examples/double_spend.rs]: Ah nice! Thank you :)

**Comment by hadjiszs** [comment]
Great!! Thanks! Are you able to push on this branch the changes you made around the incrementation of the sender sequence to make it working?
I think I checked the right option for that, thank you! :)
image

**Comment by tharvik** [comment]
> Great!! Thanks! Are you able to push on this branch the changes you made around the incrementation of the sender sequence to make it working?

Here you go, happy to help :)
You might want to rebase it on #12, as I updated `murmur` making it incompatible with older versions.

**Pull request closed without merging**